src/parse.rs

```diff
@@ -1520,41 +1521,89 @@ struct InlineEl {
 #[derive(Debug)]
 struct InlineStack {
     stack: Vec<InlineEl>,
+    // lower bounds for
+    // _, non_both, * (mod 3), ** (mod 3), *** (mod 3)
+    // for example an underscore empasis will never match
+    // with any element in the stack with index lowerbounds[0] or below
+    lower_bounds: [usize; 5],
 }
 
 impl InlineStack {
     fn new() -> InlineStack {
         InlineStack {
             stack: Vec::new(),
+            lower_bounds: [0; 5],
         }
     }
 
-    fn pop_to<'a>(&mut self, tree: &mut Tree<Item<'a>>, new_len: usize) {
-        for el in self.stack.drain(new_len..) {
+    fn pop_all<'a>(self, tree: &mut Tree<Item<'a>>) {
+        for el in self.stack {
             for i in 0..el.count {
                 tree[el.start + i].item.body = ItemBody::Text;
             }
         }
     }
 
-    fn find_match(&self, c: u8, count: usize, both: bool) -> Option<(usize, InlineEl)> {
-        self.stack
+    // both implies *, i think. because _ can never be
+    // both opener and closer
+    fn get_lowerbound(&self, c: u8, count: usize, both: bool) -> usize {
+        if c == b'_' {
+            self.lower_bounds[0]
+        } else if c == b'*' {
+            let mod3_lower = self.lower_bounds[2 + count % 3];
+            if both {
+                mod3_lower
+            } else {
+                min(mod3_lower, self.lower_bounds[1])
+            }
+        } else {
+            0
+        }
+    }
+
+    fn set_lowerbound(&mut self, c: u8, count: usize, both: bool, new_bound: usize) {
+        if c == b'_' {
+            self.lower_bounds[0] = new_bound;
+        } else if c == b'*' {
+            self.lower_bounds[2 + count % 3] = new_bound;
+            if !both {
+                self.lower_bounds[1] = new_bound;
+            }
+        }
+    }
+
+    fn find_match<'a>(&mut self, tree: &mut Tree<Item<'a>>, c: u8, count: usize, both: bool)
+        -> Option<InlineEl>
+    {
+        let lowerbound = self.get_lowerbound(c, count, both);
+        let res = self.stack[lowerbound..]
             .iter()
             .cloned()
             .enumerate()
             .rev()
             .find(|(_, el)| {
                 el.c == c && (!both && !el.both || (count + el.count) % 3 != 0)
-            })
+            });
+
+        if let Some((matching_ix, matching_el)) = res {
+            for i in (matching_ix + 1)..self.stack.len() {
+                let el = self.stack[i];
+                self.set_lowerbound(el.c, el.count, el.both, matching_ix.saturating_sub(1));
+                for i in 0..el.count {
+                    tree[el.start + i].item.body = ItemBody::Text;
+                }
+            }
+            self.stack.truncate(matching_ix);
+            Some(matching_el)
+        } else {
+            self.set_lowerbound(c, count, both, self.stack.len().saturating_sub(1));
```

Comment:
Given the above comment, would this still work without the saturating_sub(1)? I could test it, but am lazy :)

Reply:
It doesn't look like it. The stack can be empty, in which case we'd overflow..